eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ExternalDependencyExposedCheck.java

```diff
@@ -0,0 +1,201 @@
+// Licensed under the MIT License.
+
+package com.azure.tools.checkstyle.checks;
+
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.FullIdent;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ *  No external dependency exposed in the public API
+ */
+public class ExternalDependencyExposedCheck extends AbstractCheck {
+    private static final String JAVA = "java";
+    private static final String COM_AZURE = "com.azure";
+    private static final String REACTOR = "reactor";
+    private static final String DOT = ".";
+
+    private static final String EXTERNAL_DEPENDENCY_ERROR =
+        "Class ''%s'', is a class from external dependency. You should not use it as a return or parameter argument type";
+
+    private static final Set<String> validLibrarySet = new HashSet<>();
+    private static Map<String, String> classPathMap = new HashMap<>();
+
+    @Override
+    public void init() {
+        validLibrarySet.add(JAVA);
+        validLibrarySet.add(COM_AZURE);
+        validLibrarySet.add(REACTOR);
+    }
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.IMPORT,
+            TokenTypes.METHOD_DEF
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST token) {
+        switch (token.getType()) {
+            case TokenTypes.IMPORT:
+                collectAllImportedClassPath(token);
+                break;
+            case TokenTypes.METHOD_DEF:
+                checkNoExternalDependencyExposed(token);
+                break;
+            default:
+                // Checkstyle complains if there's no default block in switch
+                break;
+        }
+    }
+
+    /**
+     * Collect all imported classes into a map, key is the name of class and value is the full package path of class.
+     *
+     * @param token the IMPORT AST node
+     */
+    private void collectAllImportedClassPath(DetailAST token) {
+        final String importClassPath = FullIdent.createFullIdentBelow(token).getText();
+        final String className = importClassPath.substring(importClassPath.lastIndexOf(DOT) + 1);
+        classPathMap.put(className, importClassPath);
+    }
+
+    /**
+     * A Check for external dependency, log the error if it is an invalid external dependency.
+     *
+     * @param methodDefToken METHOD_DEF AST node
+     */
+    private void checkNoExternalDependencyExposed(DetailAST methodDefToken) {
+        final DetailAST modifiersToken = methodDefToken.findFirstToken(TokenTypes.MODIFIERS);
```

Comment:
CheckUtil has a method that does this: `public static AccessModifier getAccessModifierFromModifiersToken(DetailAST modifiersToken)`